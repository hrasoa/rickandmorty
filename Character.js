import React from 'react';
import { StyleSheet, Text, View, Image } from 'react-native';
import { connect } from 'react-redux';
import { Avatar } from 'react-native-elements';
import { singleCharacter } from './reducer';

class Character extends React.Component {
  componentDidMount() {
    const { singleCharacter, characters, navigation } = this.props;
    const id = navigation.getParam('id', null);
    if (!(id && characters[id])) {
      singleCharacter(id);
    }
  }

  render() {
    const { navigation, characters, loading, error } = this.props;
    const id = navigation.getParam('id', null);
    const character = characters[id];
    return (
      <View style={styles.container}>
        <View>
          {loading && <Text style={styles.headlines}>Loading...</Text>}
          {!loading && error && <Text style={styles.headlines}>{error}</Text>}
          {!loading && !error && character && (
            <View>
              <Avatar
                containerStyle={styles.section}
                xlarge
                rounded
                source={{ uri: character.image }}
              />
              <View style={styles.section}>
                <Text style={styles.headlines}>Name</Text>
                <Text>{character.name}</Text>
              </View>
              <View style={styles.section}>
                <Text style={styles.headlines}>Status</Text>
                <Text>{character.status}</Text>
              </View>
              <View style={styles.section}>
                <Text style={styles.headlines}>Species</Text>
                <Text>{character.species}</Text>
              </View>
              <View style={styles.section}>
                <Text style={styles.headlines}>Origin</Text>
                <Text>{character.origin.name}</Text>
              </View>
              <View style={styles.section}>
                <Text style={styles.headlines}>Gender</Text>
                <Text>{character.gender}</Text>
              </View>
            </View>
          )}
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
  section: {
    marginBottom: 24,
  },
  headlines: {
    fontWeight: 'bold',
    marginBottom: 4,
  },
});

const mapStateToProps = state => {
  const { characters, error, loading } = state;
  return {
    error,
    loading,
    characters,
  };
};

const mapDispatchToProps = {
  singleCharacter,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Character);

export { Character };
