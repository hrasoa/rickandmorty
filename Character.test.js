import React from 'react';
import renderer from 'react-test-renderer';
import { Character } from './Character';
import characters from './characters.json';

test('renders correctly', () => {
  const tree = renderer
    .create(
      <Character
        navigation={{ getParam: id => 1 }}
        characters={characters.results.reduce(
          (acc, character) => ({ ...acc, [character.id]: character }),
          {}
        )}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders loading state', () => {
  const tree = renderer
    .create(
      <Character
        navigation={{ getParam: id => 1 }}
        loading={true}
        characters={characters.results.reduce(
          (acc, character) => ({ ...acc, [character.id]: character }),
          {}
        )}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders error state', () => {
  const tree = renderer
    .create(
      <Character
        navigation={{ getParam: id => 1 }}
        error="Error ..."
        characters={characters.results.reduce(
          (acc, character) => ({ ...acc, [character.id]: character }),
          {}
        )}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
