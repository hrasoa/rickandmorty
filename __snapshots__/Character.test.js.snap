// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View>
    <View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "height": 150,
              "paddingBottom": 10,
              "paddingRight": 10,
              "paddingTop": 10,
              "width": 150,
            },
            Object {
              "borderRadius": 75,
            },
            Object {
              "marginBottom": 24,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "alignSelf": "stretch",
                "backgroundColor": "rgba(0,0,0,0.2)",
                "bottom": 0,
                "flex": 1,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              Object {
                "borderRadius": 75,
              },
              undefined,
            ]
          }
        >
          <Image
            source={
              Object {
                "uri": "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
              }
            }
            style={
              Array [
                Object {
                  "height": 150,
                  "width": 150,
                },
                Object {
                  "borderRadius": 75,
                },
                undefined,
              ]
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "marginBottom": 24,
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "marginBottom": 4,
            }
          }
        >
          Name
        </Text>
        <Text>
          Rick Sanchez
        </Text>
      </View>
      <View
        style={
          Object {
            "marginBottom": 24,
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "marginBottom": 4,
            }
          }
        >
          Status
        </Text>
        <Text>
          Alive
        </Text>
      </View>
      <View
        style={
          Object {
            "marginBottom": 24,
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "marginBottom": 4,
            }
          }
        >
          Species
        </Text>
        <Text>
          Human
        </Text>
      </View>
      <View
        style={
          Object {
            "marginBottom": 24,
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "marginBottom": 4,
            }
          }
        >
          Origin
        </Text>
        <Text>
          Earth (C-137)
        </Text>
      </View>
      <View
        style={
          Object {
            "marginBottom": 24,
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "marginBottom": 4,
            }
          }
        >
          Gender
        </Text>
        <Text>
          Male
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders error state 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "fontWeight": "bold",
          "marginBottom": 4,
        }
      }
    >
      Error ...
    </Text>
  </View>
</View>
`;

exports[`renders loading state 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "fontWeight": "bold",
          "marginBottom": 4,
        }
      }
    >
      Loading...
    </Text>
  </View>
</View>
`;
